mne/io/nedf/nedf.py

```diff
@@ -0,0 +1,193 @@
+    #   ]), (5,))
+    # ]
+
+    dt = []  # dtype for the binary data block
+    datadt = []  # dtype for a single EEG sample
+
+    headerend = header.find(b'\0')
+    if headerend == -1:
+        raise RuntimeError('End of header null not found')
+    headerxml = ElementTree.fromstring(header[:headerend])
+    nedfversion = headerxml.findtext('NEDFversion', '')
+    if nedfversion not in ['1.3', '1.4']:
+        warn('Unexpected NEDFversion, hope this works anyway')
+
+    if headerxml.findtext('AdditionalChannelStatus', 'OFF') != 'OFF':
+        raise RuntimeError('Unknown additional channel, aborting.')
+
+    n_acc = int(headerxml.findtext('NumberOfChannelsOfAccelerometer', 0))
+    if n_acc:
+        # expect one sample of u16 accelerometer data per block
+        dt.append(('acc', '>u2', (n_acc,)))
+
+    eegset = headerxml.find('EEGSettings')
+    if eegset is None:
+        raise RuntimeError('No EEG channels found')
+    nchantotal = int(_getsubnodetext(eegset, 'TotalNumberOfChannels'))
+    info['nchan'] = nchantotal
+
+    info['sfreq'] = int(_getsubnodetext(eegset, 'EEGSamplingRate'))
+    info['ch_names'] = [e.text for e in eegset.find('EEGMontage')]
+    # expect nchantotal uint24s
+    datadt.append(('eeg', 'B', (nchantotal, 3)))
+
+    if headerxml.find('STIMSettings') is not None:
+        # 2* -> two stim samples per eeg sample
+        datadt.append(('stim', 'B', (2, nchantotal, 3)))
+
+    if headerxml.findtext('stepDetails/DeviceClass', '') == 'STARSTIM':
+        warn('Found Starstim, this hasn\'t been tested extensively!')
+
+    # Trigger data: 4 bytes in newer versions, 1 byte in older versions
+    trigger_type = '>i4' if headerxml.findtext('NEDFversion') else 'B'
+    datadt.append(('trig', trigger_type))
+    # 5 data samples per block
+    dt.append(('data', np.dtype(datadt), (5,)))
+
+    date = headerxml.findtext('StepDetails/StartDate_firstEEGTimestamp', 0)
+    info['meas_date'] = datetime.fromtimestamp(int(date) / 1000, timezone.utc)
+    return info, np.dtype(dt)
+
+
+def _read_nedf_eeg(filename: str):
+    """Read header info and EEG data from an .nedf file.
+
+    Parameters
+    ----------
+    filename : str
+        Path to the .nedf file.
+
+    Returns
+    -------
+    eeg : array, shape (n_samples, n_channels)
+        Unscaled EEG data.
+    info : dict
+        Information from the file header.
+    triggers : array, shape (n_annots, 2)
+        Start samples and values of each trigger.
+    scale : float
+        Scaling factor for the EEG data.
+    """
+    # the first 10240 bytes are header in XML format, padded with NULL bytes
+    hdrlen = 10240
+    with open(filename, mode='rb') as f:
+        info, dt = _parse_nedf_header(f.read(hdrlen))
+        data = np.fromfile(f, dtype=dt)
+
+    # convert uint8-triplet -> float32
+    eeg = data['data']['eeg'] @ [1 << 16, 1 << 8, 1.]
+    eeg = eeg.reshape((-1, info['nchan']))
+    # convert sign if necessary
+    eeg[eeg > (1 << 23)] -= 1 << 24
```

Comment:
It will be pretty easy to convert this to something that properly supports `preload=True | False` if you're on board with my comment below about using STIM channels rather than annotations. Let me know if you want help with this and I can push a commit

Reply:
The reader design with the numpy dtype would be perfect for lazy loading / `mmap()`ing files, but there's not native int24 type and enumerating the annotations requires reading the data anyway so I skipped it.

Adding `preload` support would mean that all calls to `get_data()` call `_read_segment_file()`, for `preload=True` it returns the previously loaded data and for `preload=False` loads it from disk?

Reply:
> Adding preload support would mean that all calls to get_data() call _read_segment_file(), for preload=True it returns the previously loaded data and for preload=False loads it from disk?

From an implementation standpoint it means that we need to implement `_read_segment_file` properly, and never in the constructor bother reading the entire data -- the `BaseRaw` class handles using `_read_segment_file` to load all data when a user does `preload=True`. So the code here ultimately would be a bit simpler.

This is also part of why I'd like to avoid the `annotations` approach -- it requires reading all the data from disk (or at least sampling every Nth stim channel sample), which at least partially negates the benefits of `preload=False` (not needing to load the whole thing first).

> mmap()ing files

FWIW we actually have a way to do this with `preload=<string>` already in the base class, no need to implement this at the individual-reader level

Reply:
Sounds good :+1:

Does this load data and triggers in a single pass or one pass for data and triggers each?

Reply:
Single pass since it treats the trigger just like any other channel

Reply:
Do you want to take a stab at making the necessary changes or would it help if I did?

Reply:
I haven't worked with stim channels before, so I'd be glad about any help